### ISSUE TITLE
Support for Android devices (using Firefox)

### ISSUE THREAD

**Issue by BroJac5246** [ScratchAddons/ScratchAddons#6789]
### Why this would be helpful

[Mozilla recently announced support for browser extensions on the Android version of Firefox](https://blog.mozilla.org/addons/2023/08/10/prepare-your-firefox-desktop-extension-for-the-upcoming-android-release/). It is the first major browser to do so.

Scratch Addons officially supports Firefox, so bringing the extension to Android presumably wouldn't be too hard. However, the extension is not optimized for touchscreens, as brought up in #5160.

This is a pretty big deal, but the question is: will we support it? Since we really need to improve on touchscreen support, this could be good motivation, and if we do add support, this will be a game-changer for Scratch on mobile devices, particularly Android tablets.

### My suggestion

Add support for Android devices using Firefox

### Possible alternatives

 - Not add support
 - Create a new version of the extension meant for touchscreens (though that would be complicated)

### Additional context

#5160
#6512
More info: https://www.androidpolice.com/firefox-bringing-back-full-extension-support-android/

**Comment by BroJac5246** [comment]
How to enable support in the Nightly developer build:
https://blog.mozilla.org/addons/2020/09/29/expanded-extension-support-in-firefox-for-android-nightly/

**Comment by DNin01** [comment]
Note that Firefox makes up less than 5% of our user base at this time.

![](https://img.shields.io/amo/users/scratch-messaging-extension?style=flat-square&logo=firefox-browser&logoColor=white&label=users&color=FF7139) ![](https://img.shields.io/chrome-web-store/users/fbeffbjdlemaoicjdapfpikkikjoneco?style=flat-square&logo=google-chrome&logoColor=white&label=Chrome&color=E23A2E)
(There's even more users from Edge Add-ons too.)

There aren't that many [Firefox Android users](https://analytics.scratchaddons.com/scratchaddons.com?page=**%2Fwelcome%2F&browser=Firefox) either.

<details><summary>Screenshot</summary>
<p>

![Plausible Analytics for scratchaddons dot com](https://github.com/ScratchAddons/ScratchAddons/assets/106490990/ce792115-0695-4c9c-b9dd-3eb7995031f1)

</p>
</details>

[Android users across all platforms](https://analytics.scratchaddons.com/scratchaddons.com?page=**%2Fwelcome%2F&os=Android) is about 4.5k, however.

But I suppose that _could_ grow. We'll keep a lookout. There are also the touchscreen laptops.

**Comment by mybearworld** [comment]
Scratch itself doesn't support mobile devices particularly well.

**Comment by DNin01** [comment]
> Scratch itself doesn't support mobile devices particularly well.

They have OK support for tablets, but phones are way too small to do any reasonable editing.

**Comment by mybearworld** [comment]
> > Scratch itself doesn't support mobile devices particularly well.
> 
> They have OK support for tablets, but phones are way too small to do any reasonable editing.

True, I forgot about tablets

**Comment by WorldLanguages** [comment]
We're *probably* going to recommend mobile Firefox instead of Kiwi. We currently recommend Kiwi on Android as a way to use Scratch Addons on mobile.

**Comment by BroJac5246** [comment]
> > Scratch itself doesn't support mobile devices particularly well.
> 
> They have OK support for tablets, but phones are way too small to do any reasonable editing.

Scratch works great on tablets, the only real issue is that a lot of people don't include mobile support in their projects.

**Comment by BroJac5246** [comment]
> Note that Firefox makes up less than 5% of our user base at this time.

True. And it could be similar on Android. But unlike on desktop, it has no competition in the extensions space.

> ![](https://img.shields.io/amo/users/scratch-messaging-extension?style=flat-square&logo=firefox-browser&logoColor=white&label=users&color=FF7139) ![](https://img.shields.io/chrome-web-store/users/fbeffbjdlemaoicjdapfpikkikjoneco?style=flat-square&logo=google-chrome&logoColor=white&label=Chrome&color=E23A2E)
> (There's even more users from Edge Add-ons too.)
> 
> There aren't that many [Firefox Android users](https://analytics.scratchaddons.com/scratchaddons.com?page=**%2Fwelcome%2F&browser=Firefox) either.

Okay... ? How is that relevant, though? This is (currently) a nightly-exclusive developer feature, so there's no reason any Android Firefox users would have visited the website. What can we learn from that?

> <details><summary>Screenshot</summary>
> <p>
> 
> ![Plausible Analytics for scratchaddons dot com](https://github.com/ScratchAddons/ScratchAddons/assets/106490990/ce792115-0695-4c9c-b9dd-3eb7995031f1)
> 
> </p>
> </details>
> 
> [Android users across all platforms](https://analytics.scratchaddons.com/scratchaddons.com?page=**%2Fwelcome%2F&os=Android) is about 4.5k, however.

So that's all Kiwi browser? If people will install a no-name browser just to get Scratch Addons, I'd assume even more people would use Firefox, which is well-known.

> But I suppose that _could_ grow.

... If we make it grow by supporting Firefox and encouraging people to use it.

> We'll keep a lookout. There are also the touchscreen laptops.

#5160

**Comment by GrahamSH-LLK** [comment]
> Okay... ? How is that relevant, though? This is (currently) a nightly-exclusive developer feature, so there's no reason any Android Firefox users would have visited the website. What can we learn from that?
It's extremely relevant. People don't just know that we don't support firefox android, they find out by trying to install from the site. It's literally 0% of our overall Firefox traffic. This is definitely something we should do, but it is by no means a priority.

**Comment by BroJac5246** [comment]
> > Okay... ? How is that relevant, though? This is (currently) a nightly-exclusive developer feature, so there's no reason any Android Firefox users would have visited the website. What can we learn from that?
> 
> It's extremely relevant. People don't just know that we don't support Firefox android, they find out by trying to install it from the site. It's literally 0% of our overall Firefox traffic. This is definitely something we should do, but it is by no means a priority.

But everyone knows that mobile devices don't support browser extensions. So there's no reason they would even try.
> Also, they said we'd get a release date in early September. I see no release date anywhere, and it's the end of October. Also, they've been beta testing this for 3 years on Canary.

All I know is that it's a pretty explicit announcement 🤷

**Comment by GrahamSH-LLK** [comment]
> > > Okay... ? How is that relevant, though? This is (currently) a nightly-exclusive developer feature, so there's no reason any Android Firefox users would have visited the website. What can we learn from that?
> > 
> > 
> > It's extremely relevant. People don't just know that we don't support Firefox android, they find out by trying to install it from the site. It's literally 0% of our overall Firefox traffic. This is definitely something we should do, but it is by no means a priority.
> 
> But everyone knows that mobile devices don't support browser extensions. So there's no reason they would even try.
> 

I'm pretty sure that's not an "everybody knows." We get people asking all the time. 
> > Also, they said we'd get a release date in early September. I see no release date anywhere, and it's the end of October. Also, they've been beta testing this for 3 years on Canary.
> 
> All I know is that it's a pretty explicit announcement 🤷

That they've already ignored for a month.

**Comment by Secret-chest** [comment]
#5893 has mockups.

**Comment by BroJac5246** [comment]
New update:
https://blog.mozilla.org/addons/2023/11/01/is-your-extension-ready-for-firefox-for-android/

**Comment by WorldLanguages** [comment]
> First 200 Firefox for Android extension developers (to email us) get a free t-shirt!

Interesting 😆

**Comment by BroJac5246** [comment]
Looks like December 14 is the date https://www.androidpolice.com/firefox-preview-of-open-extensions-on-android/

**Comment by BroJac5246** [comment]
Can confirm this is live in stable 🎉

My phone wasn't letting me screen record for some reason, but here's my experience:

Since we don't support mobile, you have to enable desktop mode to install the extension. After that, it works fine but isn't optimized. The popup displays everything too small, and the main settings page overflows to the right. However, most addons appear to work fine— I can confirm dark mode and pause button work.

I can't explain it all without a long video, so I'd encourage you all to check it out.

I'd also like to test this on a tablet (or install the Android version on a Chromebook) to see if it works better on bigger screens, where one would be more likely to use Scratch.

Honestly, I don't think it would be too hard to add mobile support. It's pretty simple:
 1. Make all internal pages display properly (mainly CSS?).
 2. Test all addons and add a hidden "supports mobile tag" (probably not all that hard, just takes a while)
 3. On mobile versions, only display those addons.
 4. Tell Firefox we support mobile
 5. Work on making all new addons and the remaining existing addons support touchscreens (#5160) (not something that has to happen all at once)

**Comment by WorldLanguages** [comment]
We will also need to update the FAQ to recommend Firefox instead of Kiwi.

**Comment by Samq64** [comment]
> 2. Test all addons and add a hidden "supports mobile tag" (probably not all that hard, just takes a while)
> 3. On mobile versions, only display those addons.

Maybe there could be an option in more settings to show all addons and display the tag when enabled?

**Comment by Samq64** [comment]
> We will also need to update the FAQ to recommend Firefox instead of Kiwi.

https://github.com/ScratchAddons/website-v2/pull/405
I added Firefox to my misc doc updates but left Kiwi as an alternative in the FAQ.

**Comment by BroJac5246** [comment]
[Now Edge for Android is potentially adding support as well](https://www.xda-developers.com/microsoft-edge-extension-support-android/).

**Comment by WorldLanguages** [comment]
Hopefully I can focus on mobile support after we migrate the extension to Manifest V3.

If anyone is interested I suggest checking #5160 so that we can work on making most addons touchscreen-friendly. It will also benefit TurboWarp users on iPads and people using laptops with touchscreen.

**Comment by WorldLanguages** [comment]
I'm suggesting to enhance the install flow on ScratchAddons.com for Android devices here: https://github.com/ScratchAddons/website-v2/issues/446

**Comment by DNin01** [comment]
Edge for Android just added support for extensions, too.
